keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,498 @@
+
+#### Story 2
+Silently allowing workloads running on the reserved CPUs makes benchmarking infrastructure and workloads both inaccurate.
+
+### Risks and Mitigations
+
+The feature is isolated to a specific policy option `strict-cpu-reservation` under `cpuManagerPolicyOptions` and is protected by feature gate `CPUManagerPolicyAlphaOptions` or `CPUManagerPolicyBetaOptions` before the feature graduates to `Stable` i.e. enabled by default.
+
+Concern for feature impact on best-effort workloads, the workloads that do not have resource requests, is brought up.
+
+Kube-scheduler schedules pods on node allocatable (total - reserved). For best-effort pods, kube-scheduler uses default request values when scoring the nodes, see https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/util/pod_resources.go#L32 and https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/noderesources/resource_allocation.go#L123, but the scheduler does not use the default request values when fitting the nodes i.e. best-effort pods are always admitted.
+
+The concern is, when the feature graduates to `Stable`, it will be enabled by default, best-effort workloads could be starved on the node when the node runs out of non-reserved CPU cores.
+
+However, this is exactly the feature intent, best-effort workloads have no KPI requirement, they are meant to consume whatever CPU resources left on the node including starving from time to time. Best-effort workloads are not scheduled to run on the `reservedSystemCPUs` so they shall not be run on the `reservedSystemCPUs` to destablize the whole node.
+
+Nevertheless, risk mitigation, or the need for it, is under discussion.
+
+## Design Details
+
+In Kubelet, when `strict-cpu-reservation` is enabled as a policy option, we remove the reserved cores from the shared pool at the stage of calculation DefaultCPUSet and remove the `MinSharedCPUs` from the list of available cores for exclusive allocation.
+
+Feature impact can be illustrated as following:
+
+With the following Kubelet configuration:
+
+```yaml
+kind: KubeletConfiguration
+apiVersion: kubelet.config.k8s.io/v1beta1
+featureGates:
+  ...
+  CPUManagerPolicyOptions: true
+  CPUManagerPolicyAlphaOptions: true
+cpuManagerPolicy: static
+cpuManagerPolicyOptions:
+  strict-cpu-reservation: "true"
+reservedSystemCPUs: "0,32,1,33,16,48"
+...
+```
+
+When `strict-cpu-reservation` is disabled:
+```console
+# cat /var/lib/kubelet/cpu\_manager\_state
+{"policyName":"static","defaultCpuSet":"0-79","checksum":1241370203}
+```
+
+When `strict-cpu-reservation` is enabled:
+```console
+# cat /var/lib/kubelet/cpu\_manager\_state
+{"policyName":"static","defaultCpuSet":"2-15,17-31,34-47,49-63","checksum":4141502832}
+```
+
+### Risk Mitigation (Option 1)
+
+The first option is to add `numMinSharedCPUs` in `strict-cpu-reservation` option as the minimum number of CPU cores not available for exclusive allocation and expose it to Kube-scheduler for enforcement.
```

Comment:
Could `MinSharedCPUs`  be set to 0 or has a minimum value?

Reply:
The minSharedCPUs option is abandoned and left in the KEP for the record. I will update the KEP to make this very clear.